### PR TITLE
libifupdown: stop executor evaluation after an executor fails

### PR DESCRIPTION
**Pull request by nmeum**
Without this commit, all executors of a given phase will be executed
even if one of them fails. This commit adjusts handle_executors_for_phase
to stop execution if a single executor for a given phase exits with a
non-zero exit status.

This is a follow-up to 6f52157335aee39b6fc25a1dfb0a86dadd6b5c5d.

### CODE REVIEW

File: libifupdown/lifecycle.c [hunk 1]
```diff
@@ -72,15 +72,19 @@ handle_executors_for_phase(const struct lif_execute_opts *opts, char *const envp
 	if (up)
 	{
 		LIF_DICT_FOREACH(iter, &iface->vars) {
-			if (!handle_single_executor_for_phase(iter->data, opts, envp, phase, iface->ifname))
+			if (!handle_single_executor_for_phase(iter->data, opts, envp, phase, iface->ifname)) {
 				ret = false;
+				break;
```
Comment (nmeum) [inline libifupdown/lifecycle.c:77]: We could also just change this to `return false` and get rid of `ret`.

**Comment by nmeum** [comment]
I just noticed this because I have an executor which runs in the `pre-up` phase and if that executor fails it doesn't make much sense to run any other executors. For this reason, I found the current behavior somewhat surprising. However, if this is intended it could also be left as is and maybe documented in the executors man page or something.

**Comment by kaniini** [comment]
I think what @BarbarossaTM is saying is that if we stop bringing up a single interface, `ifup -a` should not stop bringing up others.

**Comment by nmeum** [comment]
> I think what @BarbarossaTM is saying is that if we stop bringing up a single interface, `ifup -a` should not stop bringing up others.

Ahh! That makes sense and is undesirable indeed. I pushed an additional commit which just causes `change_auto_interfaces` to continue bringing up interfaces when `change_interface` fails. Only a warning message is emitted in this case.

@BarbarossaTM does this address your concern?

**Comment by kaniini** [comment]
Yeah should be fine.

**Comment by nmeum** [comment]
> Yeah should be fine.

Can we merge this then or is there anything else that needs to be addressed?

File: cmd/ifupdown.c [hunk 2]
```diff
@@ -215,8 +215,10 @@ change_auto_interfaces(struct lif_dict *collection, struct lif_dict *state, stru
 		    fnmatch(opts->include_pattern, iface->ifname, 0))
 			continue;
 
-		if (!change_interface(iface, collection, state, iface->ifname, false))
-			return false;
+		if (!change_interface(iface, collection, state, iface->ifname, false)) {
+			fprintf(stderr, "%s: failed to change state for interface %s\n", argv0, iface->ifname);
+			continue;
```
Comment (BarbarossaTM) [inline cmd/ifupdown.c:220]: With that change we always return true, so I guess we should remove the `bool` return value here, as it got redundant, and update the calling code.

@kaniini what do you think?

File: libifupdown/lifecycle.c (hunk 1 above)
Comment (BarbarossaTM) [reply libifupdown/lifecycle.c:77]: Yeah, that sounds like a good idea to me, if we `break` the `ret` doesn't add much value anymore.

**Review by BarbarossaTM — CHANGES_REQUESTED** [review]
Sorry, forgot to recheck this PR, added some comments

File: libifupdown/lifecycle.c (hunk 1 above)
Comment (nmeum) [reply libifupdown/lifecycle.c]: Done, thanks!

File: cmd/ifupdown.c (hunk 2 above)
Comment (nmeum) [reply cmd/ifupdown.c:220]: Alternatively, we could also return false when one interfaces was not brought up successfully.

That is, change this to something along the lines of:

```C
bool ret = true;

LIF_DICT_FOREACH(iter, collection) {
  // …

  if (!change_interface(iface, collection, state, iface->ifname, false)) {
    ret = false;
    fprintf(…);
    continue;
  }
}

return ret;
```

Comment (kaniini) [reply cmd/ifupdown.c:220]: I think I would prefer to do what @nmeum wants to do here.

Comment (nmeum) [reply cmd/ifupdown.c:222]: Updated accordingly, please take another look.

**Comment by nmeum** [comment]
PING.

**Pull request merged**